forge/MerklePatriciaProof.sol

```diff
@@ -0,0 +1,194 @@
+     * @param encodedPath The path in the trie leading to value.
+     * @param rlpParentNodes The rlp encoded stack of nodes.
+     * @param root The root hash of the trie.
+     * @return The boolean validity of the proof.
+     */
+    function verify(
+        bytes memory value,
+        bytes memory encodedPath,
+        bytes memory rlpParentNodes,
+        bytes32 root
+    ) internal pure returns (bool) {
+        RLPReader.RLPItem memory item = RLPReader.toRlpItem(rlpParentNodes);
+        RLPReader.RLPItem[] memory parentNodes = RLPReader.toList(item);
+
+        bytes memory currentNode;
+        RLPReader.RLPItem[] memory currentNodeList;
+
+        bytes32 nodeKey = root;
+        uint256 pathPtr = 0;
+
+        bytes memory path = _getNibbleArray(encodedPath);
+        if (path.length == 0) {
+            return false;
+        }
+
+        for (uint256 i = 0; i < parentNodes.length; i++) {
+            if (pathPtr > path.length) {
+                return false;
+            }
+
+            currentNode = RLPReader.toRlpBytes(parentNodes[i]);
+            if (nodeKey != keccak256(currentNode)) {
+                return false;
+            }
+            currentNodeList = RLPReader.toList(parentNodes[i]);
+
+            if (currentNodeList.length == 17) {
+                if (pathPtr == path.length) {
+                    if (
+                        keccak256(RLPReader.toBytes(currentNodeList[16])) ==
+                        keccak256(value)
+                    ) {
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }
+
+                uint8 nextPathNibble = uint8(path[pathPtr]);
+                if (nextPathNibble > 16) {
+                    return false;
+                }
+                nodeKey = bytes32(
+                    RLPReader.toUintStrict(currentNodeList[nextPathNibble])
+                );
+                pathPtr += 1;
+            } else if (currentNodeList.length == 2) {
+                // Alternative 1 start
+                bytes memory nodeValue = RLPReader.toBytes(currentNodeList[0]);
+                uint256 traversed = _nibblesToTraverse(
+                    nodeValue,
+                    path,
+                    pathPtr
+                );
+                if (pathPtr + traversed == path.length) {
+                    //leaf node
+                    //enforce correct nibble
+                    bytes1 prefix = _getNthNibbleOfBytes(0, nodeValue);
+                    if (
+                        keccak256(RLPReader.toBytes(currentNodeList[1])) == keccak256(value) && 
+                        (prefix == bytes1(uint8(2)) || prefix == bytes1(uint8(3)))
+                    ) {
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }
+
+                //extension node
+                if (traversed == 0) {
```

Comment:
I would suggest we use the latest version here for the extension node prefix check.